### ISSUE TITLE
Suggested tweak of logic wrt visibleWhen rules.

### ISSUE THREAD

**Issue by driedl** [draperd/react-forms-processor#48]
Hi @draperd 

I had some more time to experiment with your module.

I have a need for some more complicated rules to determine if a field should be visible.

Consider 3 fields, A, B, C.

In my case, 
B is dependent on a specific A value.
C is dependent on the same A value and a specific B value. 

I have two options which both take some tweaks to your `visibleWhen' logic. The first, is a breaking change, but allows for some more complex expressions.

Here is a sample of the current 'visibleWhen' rules syntax, close to my needs:
```javascript
fieldC = { ...
       visibleWhen: [
          { id: 'A_matches_123',
            field: 'A',
            is: [123],
          },
          { id: 'B_matches_456',
            field: 'B',
            is: [456],
          },
        ],
}
```
The rules above, as I understand the code, would render the field in either case, so i would see `C` even if `A.value !== 123`. I only want `C` visible if `every` condition is true not `some`.

One option is to wrap this "visibleWhen" array in an object which contains a new flag determining what operator function to use on the array... either `some`, or `every`. 
Something like:
```javascript
visibleWhen: {
    method: 'some|every',
    rules: [ rule1, ...ruleN]
  ]
}
```

...this of course breaks your established interface, so not sure if you would like that idea.

The solution i actually use now is to leverage a logic constraint in the above use-case where I can infer:  If B is present then A must have the value C needs. I can then simply test B for the required value and my ruleset size drops to 1 and now the `some` becomes equivalent to `every`...

The problem in the existing codebase is that even when I set the `omitWhenHidden` flag on B, the value of B which makes it to the evaluateAllRules fn still has its value. You must strip the value only when calculating the final form data and not also when the hidden value hit your validation logic.

To solve this problem, I change this:
https://github.com/draperd/react-forms-processor/blob/df2349d7c75096879e9ac331c039b8acc16ce5c7/packages/core/src/utilities/utils.js#L174-L178
 to this:
```javascript
      if (rule.field && fieldsById.hasOwnProperty(rule.field)) {
        if (fieldsById[rule.field].visible) {
          return evaluateRule(rule, fieldsById[rule.field].value);
        } else {
          return evaluateRule(rule, undefined);
        }
     } else {
        return defaultResult;
      }
```

I'd like to get your opinion on these changes, or if I missed some key features, set me straight on how to accomplish this with the current code...

If you got this far, thanks for reading the long post!

**Comment by draperd** [comment]
Hi, 

Thanks for taking the time to write such a detailed explanation of the situation. I can completely understand the rationale of what you're asking and it is clearly a limitation in the current implementation.

You're definitely right in thinking that I would want to avoid making breaking changes to the existing API but I would want to make sure that it is possible to achieve commonly required use cases.

One of the goals I was aspiring to achieve was to make the JSON as human readable as possible so that in an ideal world it would read in a relatively meaningful way. I can see that the your suggestion to apply an additional attribute to toggle between "all" and "some" would definitely be a workable option, but might make it less readable.

Another option might be to add an option to include an alternative attribute `visibleWhenAny` (or `visibleWhenSome` or something better that keeps the readability`) that would allow the current structure to be maintained.

The problem with this is that it then bloats the API somewhat - and similar rules would need to be added for disabled and requirement state. 

I'm not opposed to your suggestion and it would be easy enough to test whether or not the value of `visibleWhen` is an object or an array which would allow for swapping between longhand and shorthand versions of the syntax.

What are your thoughts on these options?

**Comment by driedl** [comment]
@draperd, I like your suggested approach. `visibleWhenSome` could simply be an alias for`visibleWhen', preserving the current contract. The original could then be deprecated after all users have had a reasonable time to do the name-change refactor. 

I would also agree that the flatter structure also lends itself to being more readable.  A simple, human readable metadata syntax is one of the reasons I favoured your implementation style over other declarative form generators.  

Also, I can't see needing a large set of these `visibleWhen*` variants.  Personally, all our current needs would be solved if `visibleWhenAny` was added.

**Comment by draperd** [comment]
Cool, thanks for the feedback ... I'll try to implement these changes as soon as I can. However, I should point out that for various reasons I don't have as much time to code on this as I'd like - but I do intend to keep this project going and regularly contribute to it, but I can't make any firm commitment as to when I'll be able to add this feature.

**Comment by draperd** [comment]
As a result of #50 you should be able to do what you need to. The changes are available in version 0.0.35 -  hopefully this will enable you to implement the forms logic that you required!

**Comment by driedl** [comment]
Thanks @draperd! much appreciated.
